### PR TITLE
feat: integrate SARIF bug reporter

### PR DESCRIPTION
**Pull request by KengoTODA**
In this PR, we'll fix following known problems in SARIF reporter developed in #1184, and integrate it to SpotBugs.
refs https://github.com/spotbugs/discuss/issues/95

- [x] `version` is empty, even though SpotBugs should read it from `META-INF/MANIFEST.MF`.
- [x] `messageString` contains placeholders like `{2.givenClass}` that cannot be handled by SARIF client.
- [x] Need to find how to add location data to `frames` in `exception`.
- [x] Add `level` property to Result object
- [x] Pass [SARIF validator](https://sarifweb.azurewebsites.net/Validation)

[spotbugs.sarif.txt](https://github.com/spotbugs/spotbugs/files/4908477/spotbugs.sarif.txt) is the SARIF file generated by the SpotBugs in this branch, here is the command line that I run in the unzipped spotbugs distribution directory:

>  $JAVA_HOME/bin/java -jar lib/spotbugs.jar -textui -sarif -sourcepath /Users/kengo/GitHub/spotbugs/spotbugs/src/main/java -sourcepath /Users/kengo/GitHub/spotbugs/spotbugs/src/gui/main -output ~/Desktop/spotbugs.sarif lib/spotbugs.jar 

----

Make sure these boxes are checked before submitting your PR -- thank you!

- [x] Added an entry into `CHANGELOG.md` if you have changed SpotBugs code

[//]: # (rtdbot-start)

URL of RTD documents:
en: https://spotbugs.readthedocs.io/en/more-sarif-report/
ja: https://spotbugs.readthedocs.io/ja/more-sarif-report/

[//]: # (rtdbot-end)

### CODE REVIEW

**Comment by Kirill89** [comment]
Thank you for your PR, I use it to get SARIF output for my project but I believe one thing is missing:

```diff
--- a/spotbugs/src/main/java/edu/umd/cs/findbugs/TextUICommandLine.java
+++ b/spotbugs/src/main/java/edu/umd/cs/findbugs/TextUICommandLine.java
@@ -177,6 +177,7 @@ public class TextUICommandLine extends FindBugsCommandLine {
         addSwitch("-sortByClass", "sort warnings by class");
         addSwitchWithOptionalExtraPart("-xml", "withMessages", "XML output (optionally with messages)");
         addSwitch("-xdocs", "xdoc XML output to use with Apache Maven");
+        addSwitch("-sarif", "SARIF 2.1.0 output");
         addSwitchWithOptionalExtraPart("-html", "stylesheet", "Generate HTML output (default stylesheet is default.xsl)");
         addSwitch("-emacs", "Use emacs reporting format");
         addSwitch("-relaxed", "Relaxed reporting mode (more false positives!)");
```

Without this change I have next error

```
Exception in thread "main" java.lang.IllegalArgumentException: Unknown option: -sarif
	at edu.umd.cs.findbugs.config.CommandLine.parse(CommandLine.java:324)
	at edu.umd.cs.findbugs.config.CommandLine.expandOptionFiles(CommandLine.java:180)
	at edu.umd.cs.findbugs.FindBugs.processCommandLine(FindBugs.java:338)
	at edu.umd.cs.findbugs.FindBugs2.main(FindBugs2.java:1221)
	at edu.umd.cs.findbugs.LaunchAppropriateUI.launch(LaunchAppropriateUI.java:106)
	at edu.umd.cs.findbugs.LaunchAppropriateUI.main(LaunchAppropriateUI.java:198)
```

**Comment by KengoTODA** [comment]
@Kirill89 Cool, thanks! I've pushed the proposed fix.

File: spotbugs/src/main/java/edu/umd/cs/findbugs/sarif/Rule.java [hunk 1]
```diff
@@ -54,7 +56,7 @@ JSONObject toJSONObject() {
     }
 
     @NonNull
-    static Rule fromBugPattern(BugPattern bugPattern) {
+    static Rule fromBugPattern(BugPattern bugPattern, String formatedMessage) {
```
Comment (ThrawnCA) [inline spotbugs/src/main/java/edu/umd/cs/findbugs/sarif/Rule.java:59]: Misspelled "formatted".

File: spotbugs/src/main/java/edu/umd/cs/findbugs/sarif/SarifException.java [hunk 2]
```diff
@@ -31,22 +33,25 @@
...
         }
         List<Throwable> innerThrowables = new ArrayList<>();
         innerThrowables.add(throwable.getCause());
         innerThrowables.addAll(Arrays.asList(throwable.getSuppressed()));
-        List<SarifException> innerExceptions = innerThrowables.stream().filter(Objects::nonNull).map(SarifException::fromThrowable).collect(Collectors
-                .toList());
-        return new SarifException(throwable.getClass().getName(), message, Stack.fromThrowable(throwable), innerExceptions);
+        List<SarifException> innerExceptions = innerThrowables.stream().filter(Objects::nonNull).map(t -> fromThrowable(t, sourceFinder, baseToId))
+                .collect(Collectors
+                        .toList());
```
Comment (ThrawnCA) [inline spotbugs/src/main/java/edu/umd/cs/findbugs/sarif/SarifException.java:46]: If this is from an automatic code formatter, then it has strange priorities.

**Comment by uhafner** [comment]
Nice too see! I think I will start writing a corresponding SARIF reader on the Jenkins side as well, then I may remove the dependency to SpotBugs in my warnings plugin. Are there plans to support the SARIF format in the maven plugin?

**Pull request merged**